### ISSUE TITLE
Unread body data from prior request is interpreted as second request

### ISSUE THREAD

**Issue by jethrogb** [hyperium/hyper#1297]
In hyper 0.10 I have this simple server:

```
extern crate hyper;
use hyper::server::{Server, Request, Response};

fn hello(req: Request, res: Response) {
    res.send(format!("{:?}",req.method).as_bytes()).unwrap();
}

fn main() {
    Server::http("0.0.0.0:12346").unwrap().handle(hello).unwrap();
}
```

When sending this request stream:
```
POST / HTTP/1.1
Host: localhost:12346
User-Agent: curl/7.47.0
Accept: */*
Content-Length: 4
Content-Type: application/x-www-form-urlencoded

bodyPOST / HTTP/1.1
Host: localhost:12346
User-Agent: curl/7.47.0
Accept: */*
Content-Length: 4
Content-Type: application/x-www-form-urlencoded

body
```

The second response body is `Extension("bodyPOST")`. The unread body of the first request got read as part of the second request. You can generate these request like so:
```
curl -v -d body http://localhost:12346 http://localhost:12346
```

**Comment by seanmonstar** [comment]
This is #540, with the root being #309.

**Comment by jethrogb** [comment]
#309 is closed but the bug still exists?

**Comment by seanmonstar** [comment]
It's closed as fixed in >=0.11.

**Comment by seanmonstar** [comment]
There isn't really a way to fix this without breaking compatibility.

The problem is that the `Request` can be deconstructed, and the `HttpReader` is a public enum, which is what keeps track of whether the body has been read or not. So there isn't anywhere to add a field, or add a `Drop` impl, or anything, that could signal that the socket shouldn't be read from for a new request.

Given the effort to try to find some way to fix this without breaking 0.10.x users, and that 0.10.x isn't receiving any more development, I'm going to close this as a wontfix.
